### ISSUE TITLE
Re-work inner components structure and roles

### ISSUE THREAD

**Issue by kuzn-ilya** [kuzn-ilya/react-async-call#12]
At the moment there are several inner components supporting promise lifecycle. All of them implement conditional rendering in some scence:
1. `Running`. Its children are rendered only during promise execution.
2. `Resolved`. Its children are rendered when the promise resolved. But sub-component also renders it's children in some other tricky condition with the purpose to render accumulated data between promise-returning function calls.
3. `Rejected`. Its children are rendered only after promise rejection (but before next succesful promise resolving).
4. `Executor`. It passes `execute` method to its `children` function for running promise-returning function manually.

The main aims of creating sub-components above were:
1. to have an more declarative way for handling promise lifecycle in comparison with "classic" render props approach, but also support the "classic" one.
2. to keep sub-components API as simple as possible.

As @kitos mentioned, `Rejected` component renders its children during execution of promise and this is tricky, not clearly understandable and is subject to re-work sub-components structure.

If we state that promise can be only in three states (running, resolved and rejected) and a user should be able to identify them then we come to conclusion that 'Running', 'Resolved' and 'Rejected' components have to be in play. Only `Resolved` component should be changed: it should not render its children during promise run. It should render its children only between successful promise resolving and the next promise call.
 
But in this case we will not have a simple way (using sub-components) to render result of sequence of promise resolving (accumulating theirs results). Obviously it can be solved by introducing new sub-component (e.g. `Result`) which should render its children if promise was resolved at least once and should pass accumulated result of promise chain into its children function. So, [this example](https://codesandbox.io/s/qxxnmwnmp6) will look like:
```JSX
<PromiseRenderer params={this.state.since} mergeResult={(prevResult, currentResult) => [prevResult, ...currentResult]}>
  <table>
    // Table header render
    <tbody>
      <PromiseRenderer.Result>
        {result => result.map(item => (
          <tr key={item.id}>
            <td>{item.id}</td><td>{item.name}</td><td>{item.owner.login}</td>
          </tr>))}
      </PromiseRenderer.Result>
    </tbody>
  </table>
  <PromiseRenderer.Running>Loading...</PromiseRenderer.Running>
  <PromiseRenderer.Rejected>{status => <div>Error: {status.message}</div>}</PromiseRenderer.Rejected>
  <PromiseRenderer.Result>
    {result => <button  onClick={() => this.setState({ since: Math.max(...result.map(item => item.id)) })}>Next</button>}
  </PromiseRenderer.Result>
</PromiseRenderer>
```

To summarize:
1. `Resolved` should render its children only between successful resolving and next promise run
2. New component (e.g. `Result`) should render it's children if at least one promise-returning function was successful

@kitos, @ventrz?

**Comment by ventrz** [comment]
maybe I just didn't get you right, huh, but what if:
1. `Resolved`, `Rejected`, `Running` sub-components are maximum dumb and they will be used just to render passed children conditionally, and DON'T PASS ANY DATA to children **at all**
2. Move data layer to `Result` (or smth like this) component, which will be used to get fetched (or accumulated) data

**Comment by kuzn-ilya** [comment]
Not sure that having dumb component for promise lifecycle this will be handy, because, for example, if you catch an error you defintely should know what went wrong. So, API for theirs children should be (IMHO) as follows:
1. `Running` - supports only React children, not function. So here we have what you are want.
2. `Rejected` - supports React children as well as children function with one argument - `rejectReason`
3. `Resolved` - supports React children as well as children function with one argument - `result`. This might be suitable for simple use cases: e.g. if you don't need to collect results between promise resolving and want not to render anything while running. 

In any case, passing function as a child is up to you. If you don't need it, just pass regular react children :)

**Comment by ventrz** [comment]
Yep, missed the case with `Rejected` component. 
But still have some concerns about `Result` and `Resolved` components: both of them expose fetched data, but `Resolved` renders its children depending on Promise state.

So, my thought is just to have only one way to fetched get data.

**Comment by kuzn-ilya** [comment]
Hm... It seems like I got your point. If `Resolved` won't take a children func, it will be still possible to implement my case using composition like this:
```JSX
<PromiseRenderer.Resolved>
  <PromiseRenderer.Result>
    {result => ...}
  </PromiseRenderer.Result>
</PromiseRenderer.Resolved>
```

**Comment by kuzn-ilya** [comment]
I am closing this issue, because of changing API is not the best thing after a few month of package usage.
